[quant][graphmode] Fix a corner case in handling `if` in insert_observers

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39645 [quant][graphmode] Support quantized::batch_norm2d_relu fusion for tracing
* #39644 [quant][graphmode] Support quantization for `aten::append`
* **#39615 [quant][graphmode] Fix a corner case in handling `if` in insert_observers**

Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942110](https://our.internmc.facebook.com/intern/diff/D21942110)

jerryzh168 committed:
[quant][graphmode] Fix a corner case in handling `if` in insert_observers

Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -751,7 +751,33 @@ def forward(self, x):
+                x = self.conv(x)
+                # x is already observed
+                if self.cond:
+                    x = torch.flatten(x)
```
If this operation where instead 
```python
if self.cond:
  x = self.conv1(x)
return x
```
you would need to insert an observer  after conv1. Is this tested?

jerryzh168 replied:
not yet, I can add the test

jerryzh168 replied:
this does not work since different branches of `if` not quantized the same way

jerryzh168 replied:
there is a conflict of rules at the moment if we want to support this, we can probably think about it later

raghuramank100 replied:
Please file an issue to track this issue down the road.

raghuramank100 reviewed (commented):
A question on test coverage.

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942110](https://our.internmc.facebook.com/intern/diff/D21942110)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942110](https://our.internmc.facebook.com/intern/diff/D21942110)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942110](https://our.internmc.facebook.com/intern/diff/D21942110)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942110](https://our.internmc.facebook.com/intern/diff/D21942110)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fix a corner case in handling `if` in insert_observers"


Summary:
corner case: https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L87

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942110](https://our.internmc.facebook.com/intern/diff/D21942110)

[ghstack-poisoned]
